### PR TITLE
fix(parser,ids,viewer): harden material association handling (#1755 follow-up)

### PR DESCRIPTION
**Pull request by louistrue**
Follow-up to #1757, fixing the four adjacent defects surfaced by the #1755 adversarial sweep.

## 1. Multiple `IfcRelAssociatesMaterial` per element (was: silently lost + nondeterministic)

An element may legally carry several material associations; the single-entry `onDemandMaterialMap` kept whichever rel the parser scanned last, and the cache rebuild could pick a *different* survivor (bucket order vs file order) — same file, different answer after a cache load.

- **Deterministic primary**: the association with the **lowest rel express id** wins the map — enforced identically in the parser scan and in `rebuildOnDemandMaps` (via edge `relationshipId`, which the cache CSR columns preserve byte-exact).
- **New parser APIs** `resolveAllMaterialDefIds` / `extractAllMaterialsOnDemand`: every association, graph-backed, rel-id ordered, with `getRelated` fallback for facade graphs and map fallback for graph-less stores; occurrence-overrides-type precedence preserved. Single-value APIs delegate to element 0 so they can never drift.
- **Consumers switched to all-associations** where correctness demands it: IDS material facets (a requirement satisfied only by the second association no longer **false-fails**), `buildMaterialUsageIndex` (By Material tab + totals list the element under every material), properties panel (one MaterialCard per association), search material rules, lists/lens material names, `Pset_Material*` aggregation.
- Cache rebuild also now recognises bare `IfcMaterialLayer`/`Profile`/`Constituent` as association targets (parity with the parser).

## 2. Server-loaded stores (investigated — not fixable client-side)

Server-parsed stores carry no source bytes **and** the Rust data model emits no material entities, so material names are structurally unresolvable client-side; the By Material tab stays empty on that path. Real fix = server-side material emission (same territory as the server-parity work in #1759). Documented, out of scope here.

## 3. Partial-fraction constituent sets (was: fraction-less siblings weighted 0)

IFC allows `Fraction` on only some constituents (sum ≤ 1, remainder unallocated). Fraction-less constituents now share the remainder equally and weights normalise to 1, so they contribute to per-material totals instead of silently reporting 0 volume.

## 4. Material totals panel: type-level quantities (was: typed elements totalled 0)

When occurrences carry no populated quantity sets, the panel now falls back to the type's (`extractTypeQuantitiesOnDemand`, #1745), resolved once per type. A named-but-empty occurrence qset no longer masks populated type quantities.

## Verification

- **Fable-planned (option analysis A/B/C with full blast-radius enumeration) and Fable-adversarially-verified** (10 attack vectors: determinism parity incl. IFCX/server rel-id edge cases, map/graph/transport/cache consistency, type-expansion interaction, hot-loop perf, React key/empty-state audit, fraction algebra, IDS multi-value semantics, changeset accuracy — no correctness bug survived; all four polish findings applied).
- Tests: parser **269** (12 new: multi-association determinism in both file orders, all-defs ordering, occurrence-suppresses-type, usage-index fan-out incl. type-with-two-associations, transport round-trip, partial/full fraction weighting), IDS **290** (new: material facet sees second association), viewer **1693** (new: cache-rebuild lowest-rel-id parity under adversarial bucket order).
- CodeRabbit CLI on the diff: 1 finding (empty-qset masking) — fixed and covered above.
- **Live UI verification** (localhost, repro model with wall carrying layer set + fireproofing, doors typed with partial-fraction constituent sets + type-level Qto): By Material shows all 5 materials incl. the second association and the fraction-less constituent; totals show 0.108 m³ (0.18 type-Qto × 0.6 fraction) / 0.072 m³ (× 0.4 remainder); wall properties render both material cards.
- API surface snapshot updated (+2 exports); changeset: `@ifc-lite/parser` minor, `@ifc-lite/ids` patch.

## Behavior notes

- Models where the old last-wins rule picked a later association may report a different (now deterministic) primary material in single-value surfaces (MCP/CLI/SDK).
- Per-association apportioning means an element with a layer set *and* a fallback material contributes its volume under each association — summing all material totals can exceed model volume for such elements (documented in the panel header).

## Summary by CodeRabbit

* **New Features**
  * Viewer now surfaces all material associations end-to-end (properties panel, material cards/totals, material lists, search filtering, and IDS validation), showing one item per associated material.
  * Added public parser capabilities to resolve and retrieve all associated material definitions for an element.
* **Bug Fixes**
  * Deterministic “primary” material selection (lowest relationship express id) with cache/parsing parity, including more IFC material subtype targets.
  * Material totals, properties, and indexing now aggregate across secondary associations; type quantity fallback and constituent fraction/weight normalization are improved.
* **Tests**
  * Expanded coverage for multi-association materials and constituent fraction edge cases.
* **Chores**
  * Updated package versions.

### CODE REVIEW

File: packages/parser/src/material-resolver.ts [hunk 1]
```diff
@@ -615,26 +665,33 @@ export function buildMaterialUsageIndex(store: IfcDataStore): Map<number, Materi
...
+                            entries: [],
+                        };
+                        usage.set(leaf.id, entry);
+                    }
+                    let seen = seenPerMaterial.get(leaf.id);
+                    if (!seen) { seen = new Set(); seenPerMaterial.set(leaf.id, seen); }
+                    for (const target of targets) {
+                        if (seen.has(target)) continue;
+                        seen.add(target);
+                        entry.entries.push({ entityId: target, weight: leaf.weight });
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/parser/src/material-resolver.ts]: Fixed in 4ee03216: one entries-row per (material, element) with weights accumulating across associations that resolve to the same base material — element count stays correct, contribution no longer rel-order dependent. Regression test added (layer-set Brick + direct Brick → single row, weight 2).

File: packages/parser/src/material-resolver.ts [hunk 2]
```diff
@@ -534,12 +571,22 @@ function resolveLeaves(
...
+            // IFC allows Fraction on only SOME constituents (sum ≤ 1, the
+            // remainder unallocated). Fraction-less constituents get an equal
+            // share of that remainder instead of weight 0, so they still
+            // contribute to per-material totals; the final normalisation keeps
+            // each element's weights summing to 1 even when authored fractions
+            // are inconsistent.
+            const assigned = constituents.reduce((s, c) => s + (c.fraction ?? 0), 0);
+            const noFraction = constituents.filter((c) => c.fraction === undefined).length;
+            const remainderShare = noFraction > 0 ? Math.max(0, 1 - assigned) / noFraction : 0;
+            const provisional = constituents.map((c) => c.fraction ?? remainderShare);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/parser/src/material-resolver.ts]: Fixed in 4ee03216: Fraction parsing now keeps an authored 0.0 as an explicit zero (>= 0 instead of > 0); only ABSENT fractions share the remainder. Regression test: [0.6, 0.0, $] → [0.6, 0, 0.4].

File: apps/viewer/src/utils/spatialHierarchy.ts [hunk 3]
```diff
@@ -299,6 +299,12 @@ const MATERIAL_DEF_TYPES = new Set([
   'IFCMATERIALPROFILESETUSAGE',
   'IFCMATERIALCONSTITUENTSET',
   'IFCMATERIALLIST',
+  // Bare IfcMaterialSelect members — legal (if unusual) RelatingMaterial
+  // targets. Without these a fresh parse maps the association but a cache
+  // rebuild silently drops it (parse/cache divergence).
+  'IFCMATERIALLAYER',
+  'IFCMATERIALPROFILE',
+  'IFCMATERIALCONSTITUENT',
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/utils/spatialHierarchy.ts:309]: Fixed in e3f218e1: added the missing IFC4 subtypes (IfcMaterialLayerWithOffsets, IfcMaterialProfileWithOffsets, IfcMaterialProfileSetUsageTapering), which completes IfcMaterialSelect membership. Regression test asserts all three map through rebuildOnDemandMaps.

File: packages/parser/src/material-resolver.ts [hunk 4]
```diff
@@ -529,17 +566,29 @@ function resolveLeaves(
...
+            // fractions are inconsistent.
+            const assigned = constituents.reduce((s, c) => s + (c.fraction ?? 0), 0);
+            const noFraction = constituents.filter((c) => c.fraction === undefined).length;
+            const remainderShare = noFraction > 0 ? Math.max(0, 1 - assigned) / noFraction : 0;
+            const provisional = constituents.map((c) => c.fraction ?? remainderShare);
+            const totalWeight = provisional.reduce((s, w) => s + w, 0);
+            for (let i = 0; i < constituents.length; i++) {
+                const weight = totalWeight > 0
+                    ? provisional[i] / totalWeight
                     : (constituents.length > 0 ? 1 / constituents.length : 0);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/parser/src/material-resolver.ts]: Fixed in e3f218e1 with the suggested weight-0 semantics: an all-explicit-zero constituent set now contributes nothing instead of equal weights (totalWeight can only be 0 when every fraction is an authored 0.0, since absent fractions receive a remainder share). Regression test added.

File: packages/parser/src/material-resolver.ts [hunk 5]
```diff
@@ -61,41 +61,100 @@ export interface MaterialConstituentInfo {
...
+
+/**
+ * Resolve ALL material definition ids for an entity: every occurrence-level
+ * IfcRelAssociatesMaterial (elements may legally carry more than one), or —
+ * when the occurrence has none — the associations of its type
+ * (IfcRelDefinesByType), matching {@link extractMaterialsOnDemand}'s
+ * occurrence-overrides-type precedence. Ordered by rel express id, so
+ * index 0 is the entity's deterministic "primary" material definition.
+ */
+export function resolveAllMaterialDefIds(store: IfcDataStore, entityId: number): number[] {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/parser/src/material-resolver.ts:105]: Keeping the name: the pre-existing published export resolveMaterialDefId (singular, already on main) uses the same MaterialDef term, and this function is documented as its all-associations variant. Renaming only the plural would split the pair across two naming schemes, which hurts discoverability more than the alias helps. MaterialDefinition is itself the EXPRESS supertype name, so this reads as shorthand for an EXPRESS name rather than an invented alias.

File: packages/parser/test/material-usage-type-expansion.test.ts [hunk 6]
```diff
@@ -88,4 +88,34 @@ describe('buildMaterialUsageIndex end-to-end type expansion (#1755)', () => {
...
+
+        // #110's winning type matches the per-element precedence...
+        const winningInfo = extractMaterialsOnDemand(store, 110);
+        const winningWood = winningInfo?.constituents?.[0]?.materialName;
+        expect(winningWood === 'wood1' || winningWood === 'wood2').toBe(true);
+        // ...and the index attributes #110 ONLY under that wood; under the
+        // other wood it must not appear at all.
+        const otherWood = winningWood === 'wood1' ? 'wood2' : 'wood1';
+        expect(byName.get(winningWood!)!.entries.filter((e) => e.entityId === 110)).toHaveLength(1);
+        expect((byName.get(otherWood)?.entries ?? []).filter((e) => e.entityId === 110)).toHaveLength(0);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/parser/test/material-usage-type-expansion.test.ts]: Fixed in 888667dd — the test now asserts wood1 (rel #210, lowest express id) as the deterministic winner and that wood2 excludes #110 entirely.
